File: Examples/ObjectiveC/LineGradientExample.m
#import "LineGradientExample.h"
@import Mapbox;
@interface LineGradientExample () <MGLMapViewDelegate>
@property (nonatomic) MGLMapView *mapView;
@end

NSString *const MBXExampleLineGradient = @"LineGradientExample";

@implementation LineGradientExample

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.mapView = [[MGLMapView alloc] initWithFrame:self.view.bounds styleURL:[MGLStyle lightStyleURL]];
    self.mapView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    
    [self.mapView setCenterCoordinate:CLLocationCoordinate2DMake(38.875, -77.035)
                            zoomLevel:12
                             animated:NO];
    
    
    [self.view addSubview:self.mapView];
    
    self.mapView.delegate = self;
}

// Wait until the map is loaded before adding line layer to the map.
- (void)mapView:(MGLMapView *)mapView didFinishLoadingStyle:(MGLStyle *)style {
    [self loadGeoJSON];
}

- (void)loadGeoJSON {
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        NSString *path = [[NSBundle mainBundle] pathForResource:@"iOSLineGeoJSON" ofType:@"geojson"];
        NSData *jsonData = [NSData dataWithContentsOfFile:path];
        dispatch_async(dispatch_get_main_queue(), ^{
            [self drawPolyline:jsonData];
        });
    });
}

- (void)drawPolyline:(NSData *)geoJson {
    // Add our GeoJSON data to the map as an MGLShapeSource.
    // We can then reference this data from an MGLStyleLayer.
    MGLShape *shape = [MGLShape shapeWithData:geoJson encoding:NSUTF8StringEncoding error:nil];
    MGLSource *source = [[MGLShapeSource alloc] initWithIdentifier:@"polyline" shape:shape options: @{MGLShapeSourceOptionLineDistanceMetrics:@YES}];
    [self.mapView.style addSource:source];
    
    // Create new layer for the line.
    MGLLineStyleLayer *layer = [[MGLLineStyleLayer alloc] initWithIdentifier:@"polyline" source:source];
    
    // Set the line join and cap to a rounded end.
    layer.lineJoin = [NSExpression expressionForConstantValue:[NSValue valueWithMGLLineJoin:MGLLineJoinRound]];
    layer.lineCap = [NSExpression expressionForConstantValue:[NSValue valueWithMGLLineCap:MGLLineCapRound]];
    
    NSDictionary *stops = @{
        @0: UIColor.blueColor,
        @0.1: UIColor.purpleColor,
        @0.3: UIColor.cyanColor,
        @0.5: UIColor.greenColor,
        @0.7: UIColor.yellowColor,
        @1: UIColor.redColor
    };
    
    
    // Set the line color to a gradient
    layer.lineGradient = [NSExpression expressionWithFormat: @"mgl_interpolate:withCurveType:parameters:stops:($lineProgress, 'linear', nil, %@)", stops];
    
    // Use `NSExpression` to allow the appearance of your gradient to change and become more detailed with the map’s zoom level
    layer.lineWidth = [NSExpression expressionWithFormat:@"mgl_interpolate:withCurveType:parameters:stops:($zoomLevel, 'linear', nil, %@)",
                       @{@14: @10, @18: @20}];
    [self.mapView.style addLayer:layer];
    
}
@end

File: Examples/Swift/LineGradientExample.swift
import UIKit
import Mapbox
@objc(LineGradientExample_Swift)

class LineGradientExample_Swift: UIViewController, MGLMapViewDelegate {
    
    var mapView: MGLMapView!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view.
        mapView = MGLMapView(frame: view.bounds)
        mapView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        mapView.styleURL = MGLStyle.lightStyleURL
        mapView.delegate = self
        view.addSubview(mapView)
        mapView.setCenter(CLLocationCoordinate2D(latitude: 38.875, longitude: -77.035), zoomLevel: 12, animated: false)
    }
    
    func mapViewDidFinishLoadingMap(_ mapView: MGLMapView) {
        drawPolyline()
    }
    
    func drawPolyline() {
        // Since we know this file exists within this project, we'll force unwrap its value.
        // If this data was coming from an external server, we would want to perform
        // proper error handling for a web request/response.
        let url = URL(fileURLWithPath: Bundle.main.path(forResource: "iOSLineGeoJSON", ofType: "geojson")!)
        guard let jsonData = try? Data(contentsOf: url) else {
            preconditionFailure("Failed to parse GeoJSON file")
        }
        // Add our GeoJSON data to the map as an MGLShapeSource.
        // We can then reference this data from an MGLStyleLayer.
        // MGLMapView.style is optional, so you must guard against it not being set.
        guard let style = self.mapView.style else { return }
        guard let shapeFromGeoJSON = try? MGLShape(data: jsonData, encoding: String.Encoding.utf8.rawValue) else {
            fatalError("Could not generate MGLShape")
        }
        let source = MGLShapeSource(identifier: "polyline", shape: shapeFromGeoJSON, options: [MGLShapeSourceOption.lineDistanceMetrics: true])
        style.addSource(source)
        // Create a new style layer for the line which will determine the line's styling properties.
        let layer = MGLLineStyleLayer(identifier: "polyline", source: source)
        // Set the line join and cap to a rounded end.
        layer.lineJoin = NSExpression(forConstantValue: "round")
        layer.lineCap = NSExpression(forConstantValue: "round")
        
        let stops = [0: UIColor.blue,
                     0.1: UIColor.purple,
                     0.3: UIColor.cyan,
                     0.5: UIColor.green,
                     0.7: UIColor.yellow,
                     1: UIColor.red]
        
        // Set the line color to a gradient
        layer.lineGradient = NSExpression(format: "mgl_interpolate:withCurveType:parameters:stops:($lineProgress, 'linear', nil, %@)", stops)
        
        // Use `NSExpression` to allow the appearance of your gradient to change and become more detailed with the map’s zoom level
        layer.lineWidth = NSExpression(format: "mgl_interpolate:withCurveType:parameters:stops:($zoomLevel, 'linear', nil, %@)",
                                       [14: 10, 18: 20])
        
        style.addLayer(layer)
    }
}
